pinot-common/src/main/java/com/linkedin/pinot/common/data/FieldSpec.java

```diff
@@ -15,152 +15,209 @@
-  }
+  public void setName(String name) {
+    Preconditions.checkNotNull(name);
 
-  public FieldSpec(String name, FieldType fType, DataType dType, String delimeter) {
-    this(name, fType, dType, false, delimeter);
+    _name = name;
   }
 
-  public void setName(String name) {
-    this.name = name;
+  public String getName() {
+    return _name;
   }
 
-  public String getName() {
-    return name;
+  public void setDelimiter(String delimiter) {
+    _delimiter = delimiter;
   }
 
   public String getDelimiter() {
-    return delimiter;
+    return _delimiter;
   }
 
-  public void setDelimiter(String delimeter) {
-    delimiter = delimeter;
+  public void setFieldType(FieldType fieldType) {
+    Preconditions.checkNotNull(fieldType);
+
+    _fieldType = fieldType;
   }
 
   public FieldType getFieldType() {
-    return fieldType;
+    return _fieldType;
   }
 
-  public void setFieldType(FieldType fieldType) {
-    this.fieldType = fieldType;
+  public void setDataType(DataType dataType) {
+    Preconditions.checkNotNull(dataType);
+
+    _isDataTypeSet = true;
+    _dataType = dataType;
+
+    // Reset default null value if already set.
+    if (_isDefaultNullValueSet) {
+      updateDefaultNullValue();
+    }
   }
 
   public DataType getDataType() {
-    return dataType;
+    return _dataType;
   }
 
-  public void setDataType(DataType dataType) {
-    this.dataType = dataType;
+  public void setSingleValueField(boolean isSingleValueField) {
+    _isSingleValueField = isSingleValueField;
   }
 
   public boolean isSingleValueField() {
-    return isSingleValueField;
+    return _isSingleValueField;
   }
 
-  public void setSingleValueField(boolean isSingleValueField) {
-    this.isSingleValueField = isSingleValueField;
-  }
+  public void setDefaultNullValue(Object defaultNullValue) {
+    Preconditions.checkNotNull(defaultNullValue);
+
+    _isDefaultNullValueSet = true;
 
-  // This method is invoked via ObjectMapper, in Schema.java.
-  // Do NOT change the signature or remove it.
-  public void setDefaultNullValue(Object value) {
-    if (value != null) {
-      defaultNullValueFromString(value.toString());
+    if (_isDataTypeSet) {
+      defaultNullValueFromString(defaultNullValue.toString());
     } else {
-      this.defaultNullValue = null;
+      // Set default null value as the value passed in, and might reset when setDataType() get called.
+      _defaultNullValue = defaultNullValue;
     }
   }
 
-  // Called when trying to set a default null value from hadoop.
-  public void defaultNullValueFromString(String value) {
-    if (value == null) {
-      this.defaultNullValue = null;
-      return;
+  /**
+   * Helper function to convert the string format default null value to the correct data type.
+   *
+   * @param value string format of the default null value.
+   */
+  protected void defaultNullValueFromString(String value) {
+    DataType dataType = getDataType();
+    switch (dataType) {
```

Comment:
Agreed, but you have to watch out for existing use cases which break with your change.